File: 1615 (Maximal Network Rank)/java/Solution.java
class Solution {
    public int maximalNetworkRank(int n, int[][] roads) {
        HashMap<Integer,Integer> directlyRoad = new HashMap<>();
        int[][] adj_mat = new int[n][n];

        for(int road[]: roads){
            int a = road[0];
            int b =road[1];
            adj_mat[a][b]  = 1;
            adj_mat[b][a]  = 1;
            directlyRoad.put(a, directlyRoad.getOrDefault(a,0)+1);
            directlyRoad.put(b, directlyRoad.getOrDefault(b,0)+1);
        }


        int res = 0;
        for(int i = 0;i<n;i++){
            for(int j = i+1;j<n;j++){
                int total_network_rank = directlyRoad.getOrDefault(i,0) + directlyRoad.getOrDefault(j,0);
                if (adj_mat[i][j]==1){
                    total_network_rank--;
                } 

                res = Math.max(res, total_network_rank);
            }
        }
        return res;
    }
}

File: 1615 (Maximal Network Rank)/python/Solution.py
class Solution:
    def maximalNetworkRank(self, n: int, roads: List[List[int]]) -> int:
        
        adj_mat = [[0]*n for _ in range(n)]
        directlyRoad = collections.defaultdict(int)

        for a,b in roads:
            adj_mat[a][b] = 1
            adj_mat[b][a] = 1
            directlyRoad[a]+=1 # a->b
            directlyRoad[b]+=1 # b->a


        res = 0
        for i in range(n):
            for j in range(i+1,n):
                total_network_rank  = directlyRoad[i]+directlyRoad[j]
                if adj_mat[i][j]==1:
                    total_network_rank-=1
                res = max(res, total_network_rank)

        return res